### PR TITLE
Add support for retrieving output files from NativeLinuxExecutor

### PR DESCRIPTION
**Pull request by Fadis**
## Type of PR

- [ ] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [x] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [ ] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Importance of PR

- Importance of the issue
    - [ ] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [x] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [ ] Large (requires several days to several weeks of review)
    - [ ] Medium (requires several hours to a day of review)
    - [x] Small (trivial changes, such as build error)

## PR Overview

This change add functionality to retrieve files generated by child process executed by NativeLinuxExecutor.
This is required to run symcc using the executor.
This change introduces new member function GetOutputFiles that returns std::uint8_t ranges of the generated file contents.
As long as no output directories are specified, the behavior of executor is similar to existing. Since the default output directories are empty, no any changes are needed for existing use of the executor.

## Concerns (Optional)

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [ ] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [ ] The PR title is a summary of the changes.
- [ ] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [ ] CI is green or confirmed test run results.
- [ ] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [ ] The title of this PR summarizes the changes made by this PR properly.
- [ ] The target branch of this PR is as intended.
- [ ] The reviewer understands the issues in this PR.
- [ ] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

### CODE REVIEW

**Review by potetisensei — CHANGES_REQUESTED** [review]
Last week I didn't notice I was requested a review. I'm so sorry.
 
I think we need to discuss the specifications and functionalities of the added functions.
I feel the member functions added in this PR are too specific to SymCC currently.
Since other algorithms also use `NativeLinuxExecutor,` the specifications of return values and arguments of these functions should be carefully designed to be universal to some extent.
For example, the vectors returned from `GetOutputFiles` contain `mapped_file_t`. Maybe, some components that use `NativeLinuxExecutor` want to map some of them later, not all of them immediately. 
Also, some components may not want to remove the output directories and instead may want to append files into the directories each time they call `Run(...)`.
In other words, we need to think about how different algorithms want output directories in different ways and to what extent we can generalize them.

File: include/fuzzuf/utils/vfs/read_once.hpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ */
+#ifndef FUZZUF_INCLUDE_UTILS_VFS_READ_ONCE_HPP
+#define FUZZUF_INCLUDE_UTILS_VFS_READ_ONCE_HPP
+#include <fcntl.h>
+#include "fuzzuf/utils/filesystem.hpp"
+#include "fuzzuf/utils/map_file.hpp"
+
+namespace fuzzuf::utils::vfs::adaptor {
+
+template< typename Base >
```
Comment (potetisensei) [inline include/fuzzuf/utils/vfs/read_once.hpp]: Please describe what kind of types can be passed as `Base`. 
Since the precise conditions are too long to write, some simple explanations like "Basically, `Base` should be a type that derives from `VFS`. Additionally, it must have some extra member functions like `OpenDirectoryRecursive`. For example, LocalFilesystem meets these conditions." are enough.

File: utils/vfs/vfs.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
+#include <algorithm>
+#include <fuzzuf/utils/filesystem.hpp>
+#include <fuzzuf/utils/vfs/vfs.hpp>
+#include <fuzzuf/exceptions.hpp>
+namespace fuzzuf::utils::vfs {
+
+VFS::VFS(
+  std::vector< fs::path > &&allowed_path_
+) {
+  if( !allowed_path_.empty() ) {
```
Comment (potetisensei) [inline utils/vfs/vfs.cpp]: Probably we can make this function less nested if it processes `if( allowed_path_.empty() ) {` first.

File: utils/vfs/vfs.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        prev_len = std::distance( prev.begin(), prev.end() );
+      }
+      else if( !std::equal(
+        prev.begin(), prev.end(),
+        cur->begin(), std::next( cur->begin(), prev_len )
+      ) ) {
+        prev = *cur;
+        allowed_path->push_back( std::move( *cur ) );
+        prev_len = std::distance( prev.begin(), prev.end() );
+      }
```
Comment (potetisensei) [inline utils/vfs/vfs.cpp]: I can see what you are doing here, but definitely it should be commented and explained.
Moreover, L45-47 and L53-55 are totally the same as procedures. I think these blocks need to be combined.

File: utils/vfs/vfs.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    }
+    if( std::find_if(
+      allowed_path->begin(),
+      allowed_path->end(),
+      [rn=(*allowed_path)[ 0 ].root_name()]( const auto &p ) {
+        return p.root_name() != rn;
```
Comment (potetisensei) [inline utils/vfs/vfs.cpp]: Just to clarify, don't we have to allow paths to have different root names?(since root names are empty in *NIX in the first place, I'm just wondering why you care about this)

File: utils/vfs/vfs.cpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    allowed_path->end(),
+    requested
+  );
+  if( ge == allowed_path->begin() ) {
+    return std::nullopt;
+  }
+  const auto closest = std::prev( ge );
+  if( closest == allowed_path->end() ) {
+    return std::nullopt;
+  }
```
Comment (potetisensei) [inline utils/vfs/vfs.cpp]: I think this case can't happen, no?

Comment (Fadis) [reply utils/vfs/vfs.cpp]: fixed

File: utils/vfs/vfs.cpp (hunk 4 above)
Comment (Fadis) [reply utils/vfs/vfs.cpp]: fixed

File: utils/vfs/vfs.cpp (hunk 3 above)
Comment (Fadis) [reply utils/vfs/vfs.cpp]: fixed

File: utils/vfs/vfs.cpp (hunk 2 above)
Comment (Fadis) [reply utils/vfs/vfs.cpp]: fixed

File: include/fuzzuf/utils/vfs/read_once.hpp (hunk 1 above)
Comment (Fadis) [reply include/fuzzuf/utils/vfs/read_once.hpp]: fixed

File: utils/vfs/vfs.cpp [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    /*
+     * Otherwise, the path is same as or included by previous one.
+     * So the path should be discarded.
+     */
+  }
+  /*
+   * Although Windows has current directory for each drives, VFS provides *NIX
+   * style single current directory. As the result, accepting multiple drives
+   * causes confusing result on windows environment. Instead, multiple drives as
+   * accessible paths is forbidden in current implementation.
```
Comment (potetisensei) [inline utils/vfs/vfs.cpp]:
```suggestion
   * Although Windows has current directories for each drives, VFS provides *NIX
   * style single current directory. As the result, accepting multiple drives
   * causes confusing results on windows environment. Therefore, we forbid
   * multiple drives from being included in the current implementation.
```

File: utils/vfs/vfs.cpp [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    else if (!std::equal(prev.begin(), prev.end(), cur->begin(),
+                         std::next(cur->begin(), prev_len))) {
+      prev = *cur;
+      allowed_path->push_back(std::move(*cur));
+      prev_len = std::distance(prev.begin(), prev.end());
+    }
+    /*
+     * Otherwise, the path is same as or included by previous one.
+     * So the path should be discarded.
+     */
```
Comment (potetisensei) [inline utils/vfs/vfs.cpp]: These comments are very good. 
I think this part can be improved further by introducing a bool variable(which may be excessive though).

```suggestion
    bool should_retain;
    /*
     * Since the paths are sorted, the path shorter than previous one definitely
     * means no inclusion nor duplication. So the path should be left.
     */
    if (prev_len > cur_len) {
        should_retain = true;
    }
    /*
     * If the path doesn't contain previous path as prefix, the path is not
     * included nor duplicated. So the path should be left.
     */
    else if (!std::equal(prev.begin(), prev.end(), cur->begin(),
                         std::next(cur->begin(), prev_len))) {
        should_retain = true;
    }
    /*
     * Otherwise, the path is same as or included by previous one.
     * So the path should be discarded.
     */
     else {
         should_retain = false;
     }
     
     if (should_retain) {
        prev = *cur;
        allowed_path->push_back(std::move(*cur));
        prev_len = std::distance(prev.begin(), prev.end());
     }
```

**Review by potetisensei — APPROVED** [review]
Nice work
